libraries/SoftwareSerial/src/SoftwareSerial.cpp

```diff
@@ -0,0 +1,432 @@
+      active_listener->stopListening();
+    }
+    rx_tick_cnt = 0;
+    rx_bit_cnt = -1;
+    setSpeed(_speed);
+    active_listener = this;
+    if (!_half_duplex) {
+      active_in = this;
+    }
+    return true;
+  }
+  return false;
+}
+
+// Stop listening. Returns true if we were actually listening.
+bool SoftwareSerial::stopListening()
+{
+  if (active_listener == this) {
+    // wait for any output to complete
+    while (active_out)
+      ;
+    if (_half_duplex) {
+      setRXTX(false);
+    }
+    active_listener = nullptr;
+    active_in = nullptr;
+    // turn off ints
+    setSpeed(0);
+    return true;
+  }
+  return false;
+}
+
+inline void SoftwareSerial::setTX()
+{
+  if (_inverse_logic) {
+    LL_GPIO_ResetOutputPin(_transmitPinPort, _transmitPinNumber);
+  } else {
+    LL_GPIO_SetOutputPin(_transmitPinPort, _transmitPinNumber);
+  }
+  pinMode(_transmitPin, OUTPUT);
+}
+
+inline void SoftwareSerial::setRX()
+{
+  if (_receivePin > 0) {
+    pinMode(_receivePin, _inverse_logic ? INPUT_PULLDOWN : INPUT_PULLUP); // pullup for normal logic!
+  }
+}
+
+inline void SoftwareSerial::setRXTX(bool input)
+{
+  if (_half_duplex) {
+    if (input) {
+      if (active_in != this) {
+        setRX();
+        rx_bit_cnt = -1;
+        rx_tick_cnt = 1;
+        active_in = this;
+      }
+    } else {
+      if (active_in == this) {
+        setTX();
+        active_in = nullptr;
+      }
+    }
+  }
+}
+
+inline void SoftwareSerial::send()
+{
+  if (--tx_tick_cnt < 0) {
+    if (tx_bit_cnt++ < 10) {
+      // send data (including start and stop bits)
+      if (tx_buffer & 1) {
+        LL_GPIO_SetOutputPin(_transmitPinPort, _transmitPinNumber);
+      } else {
+        LL_GPIO_ResetOutputPin(_transmitPinPort, _transmitPinNumber);
+
+      }
+      tx_buffer >>= 1;
+      tx_tick_cnt = OVERSAMPLE - 1; // first sample sent, OVERSAMPLE - 1 more samples to go
+    } else {
+      tx_tick_cnt = 0;
+      if (_output_pending || !(_half_duplex && active_listener == this)) {
+        active_out = nullptr;
+        // When in half-duplex mode, we wait for HALFDUPLEX_SWITCH_DELAY bit-periods after the byte has
+        // been transmitted before allowing the switch to RX mode
+      } else if (tx_bit_cnt > 10 + HALFDUPLEX_SWITCH_DELAY) {
+        pinMode(_receivePin, _inverse_logic ? INPUT_PULLDOWN : INPUT_PULLUP); // pullup for normal logic!
+        rx_bit_cnt = -1;
+        rx_tick_cnt = 1;
+        active_in = this;
+        active_out = nullptr;
+      }
+    }
+  }
+}
+
+//
+// The receive routine called by the interrupt handler
+//
+inline void SoftwareSerial::recv()
+{
+  if (--rx_tick_cnt < 0) {
+    bool inbit = LL_GPIO_IsInputPinSet(_receivePinPort, _receivePinNumber) ^ _inverse_logic;
+    if (rx_bit_cnt == -1) {
+      // waiting for start bit
+      if (!inbit) {
+        // got start bit
+        rx_bit_cnt = 0;
+        rx_tick_cnt = OVERSAMPLE - 1; // first sample received, we need OVERSAMPLE - 1 more samples
```

Comment:
The former value was OVERSAMPLE+1, so here (with the new check <0) I think you need OVERSAMPLE.
My understanding is that after the start detection, we want to wait 1 bit +1/3 of bit so that sample are made not to close to the edge, but more in the middle of the bit (to be more tolerant to clock inaccuracy)

Reply:
I disagree with you analysis and I think the OVERSAMPLE+1 in the original code was an error. Let's take an example where we have a degraded signal, caused by an open-drain pull-up setup with a pull-up (a quite common setup). Now, if the pull-up and line capacitance are relatively high, we see a slowly rising signal, and a high bit coming after a low bit will be only two samples wide instead of three. When we also add the extra delay of one sample period like you suggest, we will mis-interpret the high bit. This would not be the case without the extra delay. 
At the same time, I don't see how clock inaccuracies can be remedied by a delay here, so I don't see the advantage of it.

Reply:
Also realise this: the first sample is not necessarily on the edge, but can be anywhere between the edge and 1 sample period after the edge

Reply:
I see it differently:
I agree that 1st sample is not necessarily on the edge,
it could be between the edge and the 1/3 of the bit (with OVERSAMPLE = 3).
Lets take the case that next bit after start is 1.
If the frequency of the sender is not accurate, the rising edge could arrive later than expected and thus if we continue sampling between theoretical edge of bit and 1/3 of the bit there is a risk that we sample the bit before the rising edge. 

Now If I take your example of a slowly rising edge due to high capacitance, then bit will be only 2 sample wide. But it also imply that lower state is longer than expected (otherwise baurate is far from accurate).
So if we continue sampling between theoretical edge of bit and 1/3 of the bit there is a risk that we sample the bit before the rising edge.

In both case, if we wait for a while to avoid sampling close to the theoretical edge, but more on the middle then we can tolerate drift on the frequency or slowrising edge.

![Untitled](https://user-images.githubusercontent.com/50730894/64411349-42ba9100-d08d-11e9-899d-44d027e73449.png)

Reply:
I've made a test with a NEO6-GPS some bytes are missed with `-1`
```
GPGLL,,,,,092159.00,V,N*4C$GPRMC,092200.0,V,,,,,,,060919,,,N*73
$GPVG,,,,,,,,,N*30
$GPGGA,09200.00,,,,,0,00,99.99,,,,,,*6F$GPGSA,A,1,,,,,,,,,,,,,99.99,99.9,99.99*30
$GPGSV,1,1,04,1,,,22,21,,,22,23,,,21,26,,,2*77
```
 while not if removed.

Reply:
You're right. I mistakenly assumed that we were actually oversampling here, but what is really being done is just single sampling with a higher resolution clock. And then indeed it makes sense to wait one clock period to make sure the sample is in the middle of the bit.

Reply:
So... this got me thinking. Why not do some actual oversampling? The extra overhead is minimal and it should bring us more accuracy in case of degraded signals. This will be my last bit of stubbornness on this subject, if you guys think the current code is better I will revert it :)